mopidy_mpd/network.py

```diff
@@ -119,19 +119,24 @@ def __init__(
         self.watcher = self.register_server_socket(self.server_socket.fileno())
 
     def create_server_socket(self, host, port):
-        socket_path = get_unix_socket_path(host)
-        if socket_path is not None:  # host is a path so use unix socket
-            sock = create_unix_socket()
-            sock.bind(socket_path)
-        else:
-            # ensure the port is supplied
-            if not isinstance(port, int):
-                raise TypeError(f"Expected an integer, not {port!r}")
-            sock = create_tcp_socket()
-            sock.bind((host, port))
-
-        sock.setblocking(False)
-        sock.listen(1)
+        try:  # attempt to get a socket from systemd
+            fdnames = os.environ.get("LISTEN_FDNAMES", "").split(":")
+            fd = fdnames.index("mpd") + 3  # 3 is the first systemd file handle
+            sock = socket.socket(fileno=fd)
+        except ValueError:
+            socket_path = get_unix_socket_path(host)
+            if socket_path is not None:  # host is a path so use unix socket
+                sock = create_unix_socket()
+                sock.bind(socket_path)
+            else:
+                # ensure the port is supplied
+                if not isinstance(port, int):
+                    raise TypeError(f"Expected an integer, not {port!r}")
+                sock = create_tcp_socket()
+                sock.bind((host, port))
+
+            sock.setblocking(False)
+            sock.listen(1)
         return sock
```

Comment:
I think should be refactored into a function, maybe placed right after `create_unix_socket`:

```python
def get_systemd_socket() -> Optional[socket]:
    """Attempt to get a socket from systemd."""
    try:
        fdnames = os.environ.get("LISTEN_FDNAMES", "").split(":")
        fd = fdnames.index("mpd") + 3  # 3 is the first systemd file handle
        return socket.socket(fileno=fd)
    except ValueError:
        return None
```

So it becomes something like this (untested):

```suggestion
        systemd_socket = get_systemd_socket()
        socket_path = get_unix_socket_path(host)
        
        if systemd_socket is not None:
            sock = systemd_socket
        elif socket_path is not None:
            # host is a path so use unix socket
            sock = create_unix_socket()
            sock.bind(socket_path)
        else:
            if not isinstance(port, int):
                raise TypeError(f"Expected an integer, not {port!r}")
            sock = create_tcp_socket()
            sock.bind((host, port))

        sock.setblocking(False)
        sock.listen(1)
        return sock
```

Reply:
I don't think this is any cleaner, and also your code makes an unnecessary call to `get_unix_socket_path` even if one was passed in from systemd. I don't see any reason to abstract this into a function, seeing as it will not be reused elsewhere.

Reply:
You have a valid point with `get_unix_socket_path()` being called unnecessary in my suggestion. I still think the first five lines here should go into a function to make it clearer what it does, avoiding the need for the comment here.

Here's another alternative that doesn't create any extra sockets. The `get_systemd_socket()` function is the same as before.

```suggestion
        sock = get_systemd_socket()
        if sock is not None:
            sock.setblocking(False)
            sock.listen(1)
            return sock

        socket_path = get_unix_socket_path(host)
        if socket_path is not None:
            sock = create_unix_socket()
            sock.bind(socket_path)
            sock.setblocking(False)
            sock.listen(1)
            return sock

        if not isinstance(port, int):
            raise TypeError(f"Expected an integer, not {port!r}")
        sock = create_tcp_socket()
        sock.bind((host, port))
        sock.setblocking(False)
        sock.listen(1)
        return sock
```

Once we require Python 3.10, this could become:

```py
        if sock := get_systemd_socket():
            pass
        elif socket_path := get_unix_socket_path(host)
            sock = create_unix_socket()
            sock.bind(socket_path)
        else:
            if not isinstance(port, int):
                raise TypeError(f"Expected an integer, not {port!r}")
            sock = create_tcp_socket()
            sock.bind((host, port))

        sock.setblocking(False)
        sock.listen(1)
        return sock
```

Reply:
I've incorporated your suggestion in d7a8e56983abf631ecbfea01e28094baf51ca4be.